File: BankingApp/BankingApp/ChatTableViewController.m
/*
 * Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
 * See LICENSE in the project root for license information.
 */

#import "ChatTableViewController.h"
#import "ChatCell.h"

@interface ChatTableViewController ()
@property (nonatomic, strong) NSMutableArray *dataSource;
@end

@interface ChatMessage: NSObject

@property (nonatomic, strong, nullable) NSString *chatDisplayName;
@property (nonatomic, strong, nonnull) NSString *chatMessage;
@property (nonatomic, assign) ChatSource chatSource;

@end

@implementation ChatMessage

@end

static NSString* const ParticipantCellIdentifier = @"participantCell";
static NSString* const SelfCellIdentifier = @"selfCell";
static NSString* const StatusCellIdentifier = @"statusCell";

@implementation ChatTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.dataSource = [NSMutableArray new];
    [self addStatus:@"Waiting for an agent"];
    
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 160.0;
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Actions
- (void)addStatus:(NSString *)message {
    ChatMessage *newMessage = [[ChatMessage alloc] init];
    newMessage.chatMessage = message;
    newMessage.chatSource = ChatSourceStatus;

    [self.dataSource addObject:newMessage];
    
    [self updateTable];
}

- (void)addMessage:(NSString *)message from:(NSString *)name origin:(ChatSource)source {
    ChatMessage *newMessage = [[ChatMessage alloc] init];
    newMessage.chatDisplayName = name;
    newMessage.chatMessage = message;
    newMessage.chatSource = source;
    
    [self.dataSource addObject:newMessage];
    
    [self updateTable];
}


- (void)updateTable {
    [self.tableView reloadData];
    NSIndexPath *row = [NSIndexPath indexPathForRow:self.dataSource.count - 1 inSection:0];
    [self.tableView scrollToRowAtIndexPath:row atScrollPosition:UITableViewScrollPositionTop animated:YES];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataSource.count;
}

- (void)setUpCell:(UITableViewCell *)cell withChatMessage:(ChatMessage *)message {
    if (message.chatSource == ChatSourceStatus) {
        cell.textLabel.text = message.chatMessage;
    }
    else {
        ((ChatCell *)cell).nameLabel.text = message.chatDisplayName;
        ((ChatCell *)cell).messageLabel.text = message.chatMessage;
    }
}



- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    ChatMessage *chatMessage = self.dataSource[indexPath.row];
    UITableViewCell *cell;
    
    switch (chatMessage.chatSource) {
        case ChatSourceStatus:
            cell = [tableView dequeueReusableCellWithIdentifier:StatusCellIdentifier forIndexPath:indexPath];
            break;
            
        case ChatSourceSelf:
            cell = [tableView dequeueReusableCellWithIdentifier:SelfCellIdentifier forIndexPath:indexPath];
            break;
        
        case ChatSourceParticipant:
            cell = [tableView dequeueReusableCellWithIdentifier:ParticipantCellIdentifier forIndexPath:indexPath];
            break;
            
        default:
            break;
    }
    
    [self setUpCell:cell withChatMessage:chatMessage];
    
    return cell;
        
}


/*
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"reuseIdentifier" forIndexPath:indexPath];
    
    // Configure the cell...
    
    return cell;
}
*/

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: bankingAppSwift/bankingAppSwift/ChatTableViewController.swift
//
//  ChatTableViewController.swift
//  bankingAppSwift
//
//  Created by Aasveen Kaur on 5/12/16.
//  Copyright © 2016 Aasveen Kaur. All rights reserved.
//

import UIKit

/**
 *  Enum to identify type of chat source.
 *  They can be status message, chat from self, or chat from others
 */
enum ChatSource : Int {
    case Status
    case MySelf
    case Participant
}

class ChatMessage: NSObject {
    var chatDisplayName: String = ""
    var chatMessage: String = ""
    var chatSource: ChatSource?
}

let ParticipantCellIdentifier: String = "participantCell"
let SelfCellIdentifier: String = "selfCell"
let StatusCellIdentifier: String = "statusCell"

/**
 *  Handles UI of chat table
 */
class ChatTableViewController: UITableViewController {
    
    var myDataSource:NSMutableArray = NSMutableArray()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.addStatus("Waiting for an agent")
        self.tableView.rowHeight = UITableViewAutomaticDimension
        self.tableView.estimatedRowHeight = 160.0
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
//MARK: - ACTIONS
    
    // Add a status type message
    func addStatus(message: String) {
        let newMessage: ChatMessage = ChatMessage()
        newMessage.chatMessage = message
        newMessage.chatSource = .Status
        self.myDataSource.addObject(newMessage)
        self.updateTable()
    }
    
    
    // Add a chat message
    func addMessage(message: String, from name: String, origin source: ChatSource) {
        let newMessage: ChatMessage = ChatMessage()
        newMessage.chatDisplayName = name
        newMessage.chatMessage = message
        newMessage.chatSource = source
        self.myDataSource.addObject(newMessage)
        self.updateTable()
    }
    
    func updateTable() {
        self.tableView.reloadData()
        let row: NSIndexPath = NSIndexPath(forRow: self.myDataSource.count - 1, inSection: 0)
        self.tableView.scrollToRowAtIndexPath(row, atScrollPosition: .Top, animated: true)
    }
    
    
// MARK: - TABLE VIEW DATA SOURCE
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.myDataSource.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let chatMessage: ChatMessage = self.myDataSource[indexPath.row] as! ChatMessage
        var cell: UITableViewCell
        switch (chatMessage.chatSource!) {
            
        case .Status:
            cell = tableView.dequeueReusableCellWithIdentifier(StatusCellIdentifier, forIndexPath: indexPath)
        case .MySelf:
            cell = tableView.dequeueReusableCellWithIdentifier(SelfCellIdentifier, forIndexPath: indexPath)
        case .Participant:
            cell = tableView.dequeueReusableCellWithIdentifier(ParticipantCellIdentifier, forIndexPath: indexPath)
            
        }
        
        self.setUpCell(cell, withChatMessage: chatMessage)
        return cell
    }
    
    func setUpCell(cell: UITableViewCell, withChatMessage message: ChatMessage) {
        if message.chatSource == .Status {
            cell.textLabel?.text = message.chatMessage
        }
        else {
            ((cell as! ChatCell)).nameLabel.text = message.chatDisplayName
            ((cell as! ChatCell)).messageLabel.text = message.chatMessage
        }
    }
    
}
